### PR TITLE
V2 - TypeScript/Tests improvements + Vest fix

### PR DESCRIPTION
**Pull request by jorisre**
Last step before starting new features :)

- Fix `validateAllFieldCriteria` bug in Vest resolver
- Add [preactjs/compressed-size-action](https://github.com/preactjs/compressed-size-action)
- Add tests and add consistency between tests
- Adding test components to prevent TypeScript's errors with `useForm` — related to #97 
- Improving `Resolver` types ⬇️ 

**Before**
Screen Shot 2021-01-02 at 19 20 53

**After**
Screen Shot 2021-01-02 at 19 20 32

### CODE REVIEW

**Comment by jorisre** [comment]
@bluebill1049 PR ready for review ;)

File: joi/src/__tests__/Form.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
```
Comment (bluebill1049) [inline joi/src/__tests__/Form.tsx:1]: oh man, this is great! good have to some integrated tests.

File: joi/src/__tests__/joi.ts [hunk 2]
```diff
@@ -3,31 +3,67 @@ import { joiResolver } from '..';
...
+  });
+
+  it('should return all the errors from joiResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
+    const data = {
+      password: '___',
+      email: '',
+      birthYear: 'birthYear',
+    };
+
+    const result = await joiResolver(schema)(data, undefined, true);
```
Comment (bluebill1049) [inline joi/src/__tests__/joi.ts:65]: just a quick note, going to change this `true` to an object.

```
{
  isValidateAllFieldCriteria: boolean,
  field: {
    name: string,
  },
  // potential more
}
```

maybe call it `formContext`, not sure if that's going to be too confusing, but we figure out a good name later.

File: joi/src/__tests__/joi.ts [hunk 3]
```diff
@@ -3,31 +3,67 @@ import { joiResolver } from '..';
...
+  it('should return a single error from joiResolver when validation fails', async () => {
+    const data = {
+      password: '___',
+      email: '',
+      birthYear: 'birthYear',
+    };
+
+    const result = await joiResolver(schema)(data);
+
+    expect(result).toMatchSnapshot();
```
Comment (bluebill1049) [inline joi/src/__tests__/joi.ts:55]: Minor: (don't have to change as well) maybe be specific with the error? I normally use snapshots for a successful responses.

File: joi/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import {
+  FieldValues,
+  ResolverResult,
+  UnpackNestedValue,
+} from 'react-hook-form';
+import type { AsyncValidationOptions, Schema } from 'joi';
+
+export type Resolver = <T extends Schema>(
```
Comment (bluebill1049) [inline joi/src/types.ts:8]: here is the type in the react-hook-form repo: 

I think maybe it's better to leave that at `react-hook-form` and centralize there?

```
export type Resolver<
  TFieldValues extends FieldValues = FieldValues,
  TContext extends object = object
> = (
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>> | ResolverResult<TFieldValues>;
```

File: superstruct/src/__tests__/Form.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
```
Comment (bluebill1049) [inline superstruct/src/__tests__/Form.tsx:1]: Nice!

File: superstruct/src/__tests__/superstruct.ts [hunk 6]
```diff
@@ -1,45 +1,72 @@
...
 import { superstructResolver } from '..';
 
-const Article = object({
-  id: number(),
-  title: string(),
-  isPublished: optional(boolean()),
+const Password = define('Password', (value, ctx) =>
+  value === ctx.branch[0].password);
+
+const schema = object({
```
Comment (bluebill1049) [inline superstruct/src/__tests__/superstruct.ts:21]: Honestly, I need to read those schemas doc more, but you are certainly an expert now with all the schemas. ❤️

File: superstruct/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import {
+  FieldValues,
+  ResolverResult,
+  UnpackNestedValue,
+} from 'react-hook-form';
+import { validate, Struct } from 'superstruct';
+
+type Options = Parameters<typeof validate>[2];
```
Comment (bluebill1049) [inline superstruct/src/types.ts:8]: oh, I see why you need to leave those type here, that's fine as well. probably need to include that in the doc, so user don't get confused where the type is, perhaps we should remove that type from the core repo.

File: vest/src/__tests__/Form.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
```
Comment (bluebill1049) [inline vest/src/__tests__/Form.tsx:1]: Thank you!

File: vest/src/__tests__/vest.ts [hunk 9]
```diff
@@ -31,7 +31,7 @@ const validationSuite = vest.create('form', (data: any = {}) => {
   });
 });
 
-describe('vest', () => {
+describe('zodResolver', () => {
```
Comment (bluebill1049) [inline vest/src/__tests__/vest.ts:34]: you mean `vest`?

File: vest/src/types.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import {
```
Comment (bluebill1049) [inline vest/src/types.ts:1]: 👍

File: vest/src/vest.ts [hunk 11]
```diff
@@ -37,11 +30,11 @@ const parseErrorSchema = (
   }, {});
 };
 
-export const vestResolver = <TFieldValues extends FieldValues>(
-  schema: ICreateResult,
```
Comment (bluebill1049) [inline vest/src/vest.ts:41]: oops

File: vest/src/vest.ts [hunk 12]
```diff
@@ -37,11 +30,11 @@ const parseErrorSchema = (
...
-  _: any = {},
+export const vestResolver: Resolver = (schema, _ = {}) => async (
+  values,
+  _context,
```
Comment (bluebill1049) [inline vest/src/vest.ts:35]: 🙏

File: yup/src/__tests__/Form.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
Comment (bluebill1049) [inline yup/src/__tests__/Form.tsx:1]: Awesome.

File: yup/src/types.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import {
```
Comment (bluebill1049) [inline yup/src/types.ts:1]: 👍

File: zod/src/__tests__/Form.tsx [hunk 15]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
Comment (bluebill1049) [inline zod/src/__tests__/Form.tsx:1]: 👯

**Review by bluebill1049 — APPROVED** [review]
I love those:

- Tests
- Type improvement
- Structure improvement

Awesome stuff!! 🎖

File: joi/src/types.ts (hunk 4 above)
Comment (jorisre) [reply joi/src/types.ts:8]: It's not the same type

From react-hook-form:
```ts
export type Resolver<
  TFieldValues extends FieldValues = FieldValues,
  TContext extends object = object
> = (
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>> | ResolverResult<TFieldValues>;
```

And from resolvers:
```ts
export type Resolver = <T extends Schema>(
  schema: T,
  options?: AsyncValidationOptions,
) => <TFieldValues extends FieldValues, TContext>(
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>>;
```

The generic type isn't placed at the same point. This update improve resolver's return type. But I'm open for a better idea 😉

I tried to centralized but it breaks severals things: https://github.com/react-hook-form/react-hook-form/pull/3821

File: vest/src/__tests__/vest.ts (hunk 9 above)
Comment (jorisre) [reply vest/src/__tests__/vest.ts:34]: Oops ! copy/paste

File: joi/src/__tests__/joi.ts (hunk 2 above)
Comment (jorisre) [reply joi/src/__tests__/joi.ts:65]: Ok, sound good to me. I'll change it in another PR.

Does `field` contains `ref` too ?
Do you think we have to add an option to allow users to validate only the current instead all field ? (except on submit event)
Or users have to create a custom resolver ?

Comment (bluebill1049) [reply joi/src/__tests__/joi.ts:65]: > Does field contains ref too ?

Yes.

> Do you think we have to add an option to allow users to validate only the current instead all field ? (except on submit event)

Yes, That would be awesome. I am not sure how easy it is going to be for all resolvers.

File: joi/src/types.ts (hunk 4 above)
Comment (bluebill1049) [reply joi/src/types.ts:8]: all good, I realize that after reviewing the rest of the resolvers.

**Pull request merged**